src/libstd/collections/hash/table.rs

```diff
@@ -201,23 +203,45 @@ impl<K, V, M> EmptyBucket<K, V, M> {
     pub fn table(&self) -> &M {
         &self.table
     }
-    /// Move out the reference to the table.
-    pub fn into_table(self) -> M {
-        self.table
-    }
+    // /// Move out the reference to the table.
+    // pub fn into_table(self) -> M {
+    //     self.table
+    // }
 }
 
 impl<K, V, M> Bucket<K, V, M> {
-    /// Move out the reference to the table.
-    pub fn into_table(self) -> M {
-        self.table
-    }
+    // /// Move out the reference to the table.
+    // pub fn into_table(self) -> M {
+    //     self.table
+    // }
     /// Get the raw index.
     pub fn index(&self) -> usize {
         self.idx
     }
 }
 
+impl<K, V, M> Deref for FullBucket<K, V, M> where M: Deref<Target=RawTable<K, V>> {
+    type Target = RawTable<K, V>;
+    fn deref(&self) -> &RawTable<K, V> {
+        &self.table
+    }
+}
+
+impl<K, V, M> DerefMut for FullBucket<K, V, M> where M: DerefMut<Target=RawTable<K, V>> {
```

Comment:
An alternative would be to put an equivalent but unsafe method on `Put` to fetch the table. By making the method unsafe, any violation of memory safety would still have to use unsafe code.